js/components/generate/GenerateForm.js

```diff
@@ -0,0 +1,397 @@
+import React from "react"
+
+import { Graph, populateHybridRelatives } from "../graph/Graph"
+import Disclaimer from "../common/Disclaimer"
+import { ErrorMessage } from "../common/react_modal.js.jsx"
+
+export default class GenerateForm extends React.Component {
+  constructor(props) {
+    super(props)
+    this.state = {
+      courses: "",
+      taken: "",
+      departments: "CSC, MAT, STA",
+      maxDepth: 0,
+      location: ["utsg"],
+      includeRaws: false,
+      includeGrades: false,
+      fceCount: 0,
+      showWarning: false,
+      invalidCourses: [],
+    }
+
+    this.graph = React.createRef()
+  }
+
+  setFCECount = credits => {
+    this.setState({ fceCount: credits })
+  }
+
+  incrementFCECount = credits => {
+    this.setState({ fceCount: this.state.fceCount + credits })
+  }
+
+  handleInputChange = event => {
+    const target = event.target
+    let value
+    if (target.type === "checkbox") {
+      value = target.checked
+    } else if (target.type === "select-multiple") {
+      value = Array.from(target.selectedOptions, option => option.value)
+    } else if (target.type === "number") {
+      value = parseInt(target.value)
+    } else {
+      value = target.value
+    }
+    const name = target.name
+
+    this.setState({
+      [name]: value,
+    })
+  }
+
+  handleSubmit = event => {
+    event.preventDefault()
+
+    // NOTE: Due to the aynchronous nature of state update, this does not have any effect in the running function,
+    // but plays a role in the next render
+    this.setState({ courses: this.state.courses.replaceAll(" ", "") })
```

Comment:
It's good to disallow inputs that are just whitespace, but the `setState` is not necessary here. Instead add something to the check below on line 60 (e.g. a `.trim()`)